mockingbird-compiler/src/main/kotlin/com/careem/mockingbird/MockingbirdPlugin.kt

```diff
@@ -15,116 +15,37 @@
 abstract class MockingbirdPlugin : Plugin<Project> {
 
-    private lateinit var classLoader: ClassLoaderWrapper
-    private lateinit var functionsMiner: FunctionsMiner
-    private lateinit var projectExplorer: ProjectExplorer
-    private lateinit var mockGenerator: MockGenerator
-    private val logger: Logger = Logging.getLogger(this::class.java)
+    private val mockingbirdPluginLegacyCodeGenDelegate = MockingbirdPluginLegacyCodeGenDelegate()
+    private val mockingbirdPluginKspDelegate = MockingbirdPluginKspDelegate()
 
-    private fun setupDependencies(target: Project) {
-        classLoader = ClassLoaderWrapper(projectExplorer, target)
-        functionsMiner = FunctionsMiner(classLoader)
-        mockGenerator = MockGenerator(classLoader, functionsMiner)
-    }
+    private val logger: Logger = Logging.getLogger(this::class.java)
 
     override fun apply(target: Project) {
-        val sourceSetResolver = SourceSetResolver()
-        projectExplorer = ProjectExplorer(sourceSetResolver)
-        try {
-            target.extensions.add<MockingbirdPluginExtension>(
-                EXTENSION_NAME, MockingbirdPluginExtensionImpl(target.objects)
-            )
-
-            target.gradle.projectsEvaluated {
-                val generateMocksTask = target.task(GradleTasks.GENERATE_MOCKS) {
-                    dependsOn(target.tasks.getByName(GradleTasks.JVM_JAR))
-                    doFirst {
-                        val outputDir = targetOutputDir(target)
-                        outputDir.deleteRecursively()
-                    }
-                    doLast {
-                        generateMocks(target)
-                    }
-                }
-
-                target.tasks.forEach { task ->
-                    if (task.name.contains("Test") && (task is KotlinCompile<*>)) {
-                        task.dependsOn(generateMocksTask)
-                    }
-                }
-
-                configureSourceSets(target)
-
-                projectExplorer.visitRootProject(target.rootProject)
-                // Add test dependencies for classes that need to be mocked
-                val dependencySet = projectExplorer.explore(target)
-                target.extensions.getByType(KotlinMultiplatformExtension::class.java).run {
-                    sourceSets.getByName("commonTest") {
-                        dependencies {
-                            dependencySet.forEach { implementation(it) }
-                        }
-                    }
-                }
-            }
-
-        } catch (e: Exception) {
-            // Useful to debug
-            e.printStackTrace()
-            throw e
+        target.extensions.add<MockingbirdPluginExtension>(
+            EXTENSION_NAME, MockingbirdPluginExtensionImpl(target.objects)
+        )
+        if (legacyCodeGenRequired(target)) {
+            mockingbirdPluginLegacyCodeGenDelegate.apply(target)
         }
+        mockingbirdPluginKspDelegate.apply(target)
```

Comment:
just to confirm, this is not either/or case, right?

You want to apply ksp `mockingbirdPluginKspDelegate.apply(target)` even if legacy is required, right?

Reply:
yes, so the idea is to allow both on the same project, to give an example suppose the legacy mode is used everywhere and you add a new test with a new dependency to mock, In that test if you apply the ksp plugin you can mock the new dependency using the new approach ( @Mock ) keeping backward compatibility with the old code.